### ISSUE TITLE
Wrappers

### ISSUE THREAD

**Issue by zsunberg** [JuliaReinforcementLearning/CommonRLInterface.jl#40]
I am planning to create a few utilities for wrapping environments:

1. The first is an abstract base class, `AbstractEnvWrapper` to make defining your own wrappers easy. The user will write the following code
    ```julia
    struct MyWrapper{E} <: AbstractEnvWrapper
        env::E
    end
    
    CommonRLWrappers.wrapped_env(w::MyWrapper) = w.env
    ```
    All of the `CommonRLInterface` functions will be automatically forwarded to the environment returned by `wrapped_env`, and `provided` will be handled appropriately, then users can modify specific parts of the behavior by inheriting most of it.
2. The second is a wrapper type, `QuickEnvWrapper` to be used when someone wants to quickly override some behavior with keyword arguments. For example, if you wanted to provide a discrete action space for an environment that normally has a continuous one, you might do something like:
    ```julia
    w = QuickEnvWrapper(env, actions=[-1.0, 0.0, 1.0])
    ```
    The keyword arguments could either be static objects or functions, similar to QuickPOMDPs.jl
    
Does anyone have any comments?

One question is what the `wrapped_env` function should be called. One elegant, but potentially confusing  option could be to use `Base.getindex(w::MyWrapper) = w.env`. This would allow access to the wrapped environment with `w[]`.

**Comment by findmyway** [comment]
> One question is what the wrapped_env function should be called. One elegant, but potentially confusing option could be to use Base.getindex(w::MyWrapper) = w.env. This would allow access to the wrapped environment with w[].

This is intuitive to me. My question is that, if a wrapper is wrapped inside another wrapper, `w[]` will return the innermost `env` instead of `wrapper`, right?

**Comment by zsunberg** [comment]
@findmyway  This is an interesting thought. I had assumed the opposite - it would return the next level down. Can you describe a case where someone would want the innermost env?

In my mind, only the person who creates `AbstractEnvWrapper` and the person who creates `MyWrapper <: AbstractEnvWrapper` will need to deal with this function, neither of whom should need access to the innermost env. Maybe this will be clear once I have an implementation :)

**Comment by zsunberg** [comment]
Shoot... just ran into a roadblock. I wanted to do `abstract type AbstractWrapper <: AbstractEnv end`, but this is not good because we also have `AbstractMarkovEnv` and `AbstractZeroSumEnv`.

My first thought is to get rid of `AbstractMarkovEnv` and `AbstractZeroSumEnv`.

Will need to think about this more - any input is helpful!

**Comment by zsunberg** [comment]
The other obvious solution is to have `AbstractMarkovWrapper <: AbstractMarkovEnv` and `AbstractZeroSumWrapper <: AbstractZeroSumEnv`.

**Comment by findmyway** [comment]
>  it would return the next level down. Can you describe a case where someone would want the innermost env?

I see. I think this is also fine.

A wrapper may (or may not) depend on the specific type of environment. So I prefer to get rid of `AbstractMarkovEnv` and `AbstractZeroSumEnv` first. And when it's really necessary to distinguish the innermost env, we can still unwrap it and extract the type info.

**Comment by zsunberg** [comment]
Hmmm.... I definitely see what you're saying, but doesn't seem like wrappers should totally dictate our decisions about the type hierarchy. The question we need to revisit is whether there should be different types for games and MDPs/POMDPs. I know much less about games than (PO)MDPs, but I will think about it tomorrow.

Is OpenSpiel the best interface to look at as a model of a game interface or are there other packages?

My initial thought was that multiplayer games are different enough for MDPs/POMDPs that they could use their own type, especially because `act!` might take multiple actions and return multiple rewards. But there may be other ways than a type hierarchy to express this. If anyone else who is more interested in games (e.g. maybe @jonathan-laurent, it is much appreciated).

Another option is to punt on this for now. We could basically just focus on environments where one player plays at a time and deal with games where multiple players play simultaneously later.

**Comment by findmyway** [comment]
> Another option is to punt on this for now. We could basically just focus on environments where one player plays at a time and deal with games where multiple players play simultaneously later.

Yes, agree. And I think even for multiple players (simultaneously or not), we can still treat it as a single player scenario (a player manager). It's just that this player is smart enough to select the appropriate sub-policies and assign credits.

**Comment by zsunberg** [comment]
Ok, I think we will be able to figure out a way to handle multiagent envs in the future with only one AbstractEnv

**Comment by zsunberg** [comment]
Closed with #42
